### PR TITLE
Allow disabling DNSSEC with cdflag

### PR DESCRIPTION
**Pull request by Dhravya**

### CODE REVIEW

**Comment by Dhravya** [comment]
Hey @MattIPv4 can you please review this and let me know if any changes need to be made? I was unable to test it

File: src/utils/dig.js [hunk 1]
```diff
@@ -28,15 +28,15 @@ export const validateDomain = (input, response) => {
     };
 };
 
-export const handleDig = async ({ domain, types, short, provider }) => {
+export const handleDig = async ({ domain, types, short, cdflag ,provider }) => {
     // Make the DNS queries
     const results = await Promise.all(types.map(type =>
         performLookupWithCache(domain, type, provider.doh).then(data => ({ type, data }))));
```
Comment (MattIPv4) [inline src/utils/dig.js]: You need to pass the cdflag into the actual DoH call, so that we can actually disable it -- everything you've done so far is just visual.

File: src/utils/dig.js [hunk 2]
```diff
@@ -72,12 +72,12 @@ export const handleDig = async ({ domain, types, short, provider }) => {
     };
 
     // Convert results to an embed
-    return results.map(({ type, data }) => createEmbed(`${type} records`, present(type, data), 'diggy diggy hole'));
+    return results.map(({ type, data }) => createEmbed(`${type} records`, present(type, data), 'diggy diggy hole :warning: cd bit set for request, DNSSEC validation disabled'));
```
Comment (MattIPv4) [inline src/utils/dig.js]: As I noted in the original issue, we should only be adding the warning if the cdflag is actively being set. Showing this warning all the time would be incorrect and misleading

File: src/utils/dig.js [hunk 3]
```diff
@@ -72,12 +72,12 @@ export const handleDig = async ({ domain, types, short, provider }) => {
...
 };
 
 export const parseEmbed = embed => {
     // Match the domain name, type and if the short format was requested
-    const descMatch = embed.description.match(/^`(\S+) (\S+) @(\S+) \+noall \+answer( \+short)?`\n/);
+    const descMatch = embed.description.match(/^`(\S+) (\S+) @(\S+) \+noall \+answer( \+short \+cdflag)?`\n/);
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
    const descMatch = embed.description.match(/^`(\S+) (\S+) @(\S+) \+noall \+answer( \+short)?( \+cdflag)?`\n/);
```
You need a separate capture group here, otherwise you're never going to detect it being set.

**Review by MattIPv4 — CHANGES_REQUESTED** [review]
Thank you for the work so far, looking good but you're missing the actual logic changes needed within the DNS lookup to set the flag.

I recommend following the steps in https://github.com/MattIPv4/DNS-over-Discord#development so that you can test this functionality is actually behaving as you'd expect locally.

(For testing, `dnssec-failed.org` should result in a servfail normally, but should return A records when the cdflag is set.)

File: src/utils/dig.js (hunk 2 above)
Comment (MattIPv4) [reply src/utils/dig.js]: I should also note that I don't think the footer text is the appropriate place for this, it should be in the main body of the embed itself, after the codeblock that contains the results.

Comment (Dhravya) [reply src/utils/dig.js]: Alright!

File: src/utils/dig.js (hunk 1 above)
Comment (Dhravya) [reply src/utils/dig.js]: OOPS my bad. I have a lot to learn from this haha

Comment (Dhravya) [reply src/utils/dig.js]: Hey @MattIPv4 so, I need to add one more parameter, `cdflag` to `performLookupWithCache` and `performLookup` right?

Comment (MattIPv4) [reply src/utils/dig.js]: Yup, and then pass it into the DoH call itself.

https://github.com/MattIPv4/DNS-over-Discord/blob/8bbb3f6a5351e6ce22ec56d4e1449686fd71c918/src/utils/dns.js#L94-L103

https://github.com/MattIPv4/DNS-over-Discord/blob/8bbb3f6a5351e6ce22ec56d4e1449686fd71c918/src/utils/dns.js#L78-L81

**Comment by MattIPv4** [comment]
👋 I'm not sure what exactly your last commit did, but you appear to have caused a diff across every single file in the repo now? Have you changed the line endings on everything perhaps?

**Comment by Dhravya** [comment]
That's weird - I've not changed anything manually, except for the `dns.js` file

**Comment by MattIPv4** [comment]
If you're using a Windows system, make sure git is configured to use lf endings, not crlf endings. It should warn you about this though when you staged/committed this massive change.

**Comment by Dhravya** [comment]
Oh - I got it now. Any way to undo it?

**Comment by MattIPv4** [comment]
If you've updated git to use lf endings, I'd expect it'd now let you commit again to revert it all -- or you could just drop the last commit and redo the change.

**Comment by Dhravya** [comment]
HELL YES! Done 
image

**Comment by Dhravya** [comment]
Ok I was testing if everything works fine and I noticed a mistake i made

So I followed the cloudflare docs (https://developers.cloudflare.com/1.1.1.1/encryption/dns-over-https/) and added the `cd` bit to the JSON and DNS lookup snippets, but this causes an error when using google's DNS lookup (8.8.8.8)


```
SyntaxError: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
    at async performLookup (worker.js:10086:42)
    at async Promise.all (index 0)
    at async handleDig (worker.js:14844:21)
    at async worker.js:17271:35 at line 0, col 0
 ```

**Comment by Dhravya** [comment]
Wait.... it should still work with 8.8.8.8, that's odd 
image

**Comment by Dhravya** [comment]
dns.google is returning Error 400 (Bad Request)!!

**Comment by MattIPv4** [comment]
Before I can start reviewing this, you're going to need to revert the line ending changes that you've made to every file here

**Comment by Dhravya** [comment]
Ok! I'll revert the commit and do the same changes again

**Comment by Dhravya** [comment]
Done! @MattIPv4

File: src/utils/dns.js [hunk 4]
```diff
@@ -53,53 +53,65 @@ const processAnswer = (type, answer) => {
...
+                    entry.data = `${flags} ${tag} '${value}'`;
                     continue;
                 }
 
                 // Normal hex data
-                entry.data = words.map(part => String.fromCharCode(parseInt(part, 16))).join('').trim();
+                entry.data = words
+                    .map((part) => String.fromCharCode(parseInt(part, 16)))
+                    .join('')
+                    .trim();
```
Comment (MattIPv4) [inline src/utils/dns.js]: Can we revert the unrelated changes, please?

File: src/utils/dns.js [hunk 5]
```diff
@@ -53,53 +53,65 @@ const processAnswer = (type, answer) => {
...
     query.searchParams.set('type', type.toLowerCase());
+    query.searchParams.set('cd', cdflag);
 
     // Make our request
     return fetch(query.href, {
         headers: {
             Accept: 'application/dns-json',
         },
-    }).then(res => res.json());
+    }).then((res) => res.json());
```
Comment (MattIPv4) [inline src/utils/dns.js]:
```suggestion
    }).then(res => res.json());
```
Unrelated change

File: src/utils/dns.js [hunk 6]
```diff
@@ -53,53 +53,65 @@ const processAnswer = (type, answer) => {
...
     // Build the query packet
     const packet = encode({
         type: 'query',
         id: randInt(1, 65534),
-        flags: RECURSION_DESIRED,
-        questions: [ {
-            name: domain,
-            type,
-        } ],
+        flags: RECURSION_DESIRED | (cdflag && CHECKING_DISABLED),
```
Comment (MattIPv4) [inline src/utils/dns.js]:
```suggestion
        flags: RECURSION_DESIRED | (cdflag ? CHECKING_DISABLED : 0),
```
Just a nitpick so we aren't relying on `false` type-casting to `0`

File: src/utils/dns.js [hunk 7]
```diff
@@ -53,53 +53,65 @@ const processAnswer = (type, answer) => {
...
+        questions: [
+            {
+                name: domain,
+                type,
+            },
+        ],
```
Comment (MattIPv4) [inline src/utils/dns.js]:
```suggestion
        questions: [ {
            name: domain,
            type,
        } ],
```
Unrelated change

File: src/utils/dns.js [hunk 8]
```diff
@@ -111,30 +123,34 @@ const performLookupDns = async (domain, type, endpoint) => {
...
-    }).then(res => res.arrayBuffer()).then(data => {
-        const dec = decode(Buffer.from(data));
-        return {
-            Status: toRcode(dec.rcode),
-            Question: dec.questions,
-            Answer: dec.answers,
-        };
-    });
+    })
+        .then((res) => res.arrayBuffer())
```
Comment (MattIPv4) [inline src/utils/dns.js]:
```suggestion
        .then(res => res.arrayBuffer())
```

File: src/utils/dns.js [hunk 9]
```diff
@@ -111,30 +123,34 @@ const performLookupDns = async (domain, type, endpoint) => {
...
+        .then((data) => {
```
Comment (MattIPv4) [inline src/utils/dns.js]:
```suggestion
        .then(data => {
```

File: wrangler.toml [hunk 10]
```diff
@@ -1,30 +1,30 @@
...
+    { binding = "CACHE", id = "a4e14492f7624d39810022793969428b" }
+]
+
+[env.production]
+zone_id = "63feb9b9ee057b117f7c833928b0d21f"
+workers_dev = false
+route = "dns-over-discord.v4.wtf/*"
+kv_namespaces = [
+    { binding = "CACHE", id = "ac1c24f2c9594a97aeb7dd30377b45ef" }
+]
```
Comment (MattIPv4) [inline wrangler.toml]: Looks like you've still got the line ending issue here

File: src/utils/dig.js [hunk 11]
```diff
@@ -28,23 +28,25 @@ export const validateDomain = (input, response) => {
...
         // Error message
         if (typeof data === 'object' && data.message)
             return `${digCmd}\n${data.message}`;
 
         // No results
         if (typeof data !== 'object' || !Array.isArray(data.answer) || data.answer.length === 0)
-            return `${digCmd}\nNo records found`;
+            return `${digCmd}\nNo records found \n${(cdflag
+                ? ":warning: cd bit set for request, DNSSEC validation disabled"
+                : "")}`;
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
            return `${digCmd}\nNo records found${(cdflag
                ? "\n\n:warning: cd bit set for request, DNSSEC validation disabled"
                : "")}`;
```

File: src/utils/dig.js [hunk 12]
```diff
@@ -61,14 +63,14 @@ export const handleDig = async ({ domain, types, short, provider }) => {
...
+        // Keep adding rows until we reach Discord 4096 char limit ()
         for (const row of sourceRows) {
-            if (output([...finalRows, row]).length > 4096) break;
+            if (output([...finalRows, row]).length > (cdflag ? 4000 : 4096)) break;
             finalRows.push(row);
         }
 
         // Render and return final rows
-        return output(finalRows);
+        return  output(finalRows) + (cdflag ? "\n\n:warning: cd bit set for request, DNSSEC validation disabled" : "") ;
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
        return output(finalRows) + (cdflag ? "\n\n:warning: cd bit set for request, DNSSEC validation disabled" : "") ;
```

**Review by MattIPv4 — CHANGES_REQUESTED** [review]

File: src/utils/dns.js (hunk 4 above)
Comment (Dhravya) [reply src/utils/dns.js]: alright. Sorry about that, automatically formatted stuff

Comment (Dhravya) [reply src/utils/dns.js]: Wait, how do i *revert* it? By making another commit right?

Comment (MattIPv4) [reply src/utils/dns.js]: Yes, you can just commit the old version

Comment (Dhravya) [reply src/utils/dns.js]: cool, done

**Comment by Dhravya** [comment]
Hey @MattIPv4 I think all the changes are done (soo sorry for all the trouble)

File: wrangler.toml [hunk 13]
```diff
@@ -1,30 +1,30 @@
...
```
Comment (MattIPv4) [inline wrangler.toml]: This file still has an unexpected diff

File: src/utils/dns.js [hunk 14]
```diff
@@ -150,14 +155,14 @@ const performLookup = async (domain, type, endpoint) => {
...
+export const performLookupWithCache = (domain, type, endpoint, cdflag) => cache(
     performLookup,
-    [ domain, type, endpoint ],
+    [ domain, type, endpoint, cdflag ],
     `dns-${domain}-${type}-${endpoint.endpoint}`,
     Number(process.env.CACHE_DNS_TTL) || 10,
 );
 
-// Ordered by "popularity", dig command offers the first 25, multi-dig supports all
+// Ordered by 'popularity', dig command offers the first 25, multi-dig supports all
```
Comment (MattIPv4) [inline src/utils/dns.js]:
```suggestion
// Ordered by "popularity", dig command offers the first 25, multi-dig supports all
```
Unrelated change

File: src/utils/dns.js [hunk 15]
```diff
@@ -90,12 +91,12 @@ const performLookupJson = async (domain, type, endpoint) => {
...
 const randInt = (min, max) => Math.floor(Math.random() * (max - min + 1)) + min;
 
-const performLookupDns = async (domain, type, endpoint) => {
+const performLookupDns = async (domain, type, endpoint, cdflag) => {
     // Build the query packet
     const packet = encode({
         type: 'query',
         id: randInt(1, 65534),
-        flags: RECURSION_DESIRED,
+        flags: RECURSION_DESIRED | (cdflag ? CHECKING_DISABLED : 0),
```
Comment (MattIPv4) [inline src/utils/dns.js:99]: `CHECKING_DISABLED` needs to be imported?

File: src/utils/dig.js [hunk 16]
```diff
@@ -28,23 +28,25 @@ export const validateDomain = (input, response) => {
...
 
         // Error message
         if (typeof data === 'object' && data.message)
             return `${digCmd}\n${data.message}`;
 
         // No results
         if (typeof data !== 'object' || !Array.isArray(data.answer) || data.answer.length === 0)
-            return `${digCmd}\nNo records found`;
+            return `${digCmd}\nNo records found${(cdflag
+                ? "\n\n:warning: cd bit set for request, DNSSEC validation disabled"
```
Comment (MattIPv4) [inline src/utils/dig.js]: Can we move this warning message to a const at the top of the file as we're using it in multiple places?

**Review by MattIPv4 — CHANGES_REQUESTED** [review]
There are a few eslint errors that need resolving

File: src/utils/dig.js (hunk 16 above)
Comment (Dhravya) [reply src/utils/dig.js]: ok yeah, makes sense

File: wrangler.toml (hunk 13 above)
Comment (Dhravya) [reply wrangler.toml]: Aaahhh i'm unable to fix this

Comment (MattIPv4) [reply wrangler.toml]: Why? Should just be a case of resetting the file to as it is in master? `git checkout master -- wrangler.toml`

File: src/utils/dig.js [hunk 17]
```diff
@@ -5,6 +5,8 @@ import providers from './providers.js';
 import { presentTable } from './table.js';
 import { createEmbed } from './embed.js';
 
+const DNSSEC_DISABLED_WARNING_MESSAGE = ':warning: cd bit set for request, DNSSEC validation disabled';
```
Comment (MattIPv4) [inline src/utils/dig.js:8]: Can we include the newlines in this, rather than needing to add them in everywhere later on?

File: src/utils/dig.js [hunk 18]
```diff
@@ -61,14 +65,14 @@ export const handleDig = async ({ domain, types, short, provider }) => {
             return `${digCmd}\`\`\`\n${rowsStr}${truncStr}\n\`\`\``;
         };
 
-        // Keep adding rows until we reach Discord 4096 char limit
+        // Keep adding rows until we reach Discord 4096 char limit ()
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
        // Keep adding rows until we reach Discord 4096 char limit
```

File: src/utils/dig.js [hunk 19]
```diff
@@ -61,14 +65,14 @@ export const handleDig = async ({ domain, types, short, provider }) => {
...
         for (const row of sourceRows) {
-            if (output([...finalRows, row]).length > 4096) break;
+            if (output([...finalRows, row]).length > (cdflag ? 4000 : 4096)) break;
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
            if (output([...finalRows, row]).length > (4096 - (cdflag ? DNSSEC_DISABLED_WARNING_MESSAGE.length : 0)) break;
```
(Might be worth moving `(4096 - (cdflag ? DNSSEC_DISABLED_WARNING_MESSAGE.length : 0)` to a max length const before this loop)

**Review by MattIPv4 — CHANGES_REQUESTED** [review]

**Comment by Dhravya** [comment]
Fixed eslint errors

File: src/utils/dig.js [hunk 20]
```diff
@@ -28,23 +30,27 @@ export const validateDomain = (input, response) => {
...
 };
 
-export const handleDig = async ({ domain, types, short, provider }) => {
+export const handleDig = async ({ domain, types, short, cdflag, provider }) => {
     // Make the DNS queries
     const results = await Promise.all(types.map(type =>
-        performLookupWithCache(domain, type, provider.doh).then(data => ({ type, data }))));
+        performLookupWithCache(domain, type, provider.doh, cdflag).then(data => ({ type, data }))));
+
+    const max_length = 4096 - (cdflag ? DNSSEC_DISABLED_WARNING_MESSAGE.length : 0);
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
    const maxLength = 4096 - (cdflag ? DNSSEC_DISABLED_WARNING_MESSAGE.length : 0);
```
JavaScript conventionally uses camelCase, not snake_case

File: src/utils/dns.js [hunk 21]
```diff
@@ -74,11 +74,12 @@ const processAnswer = (type, answer) => {
     return answer;
 };
 
-const performLookupJson = async (domain, type, endpoint) => {
+const performLookupJson = async (domain, type, endpoint, cdflag) => {
     // Build the query URL
     const query = new URL(endpoint.endpoint);
     query.searchParams.set('name', domain);
     query.searchParams.set('type', type.toLowerCase());
+    query.searchParams.set('cd', cdflag);
```
Comment (MattIPv4) [inline src/utils/dns.js]:
```suggestion
    query.searchParams.set('cd', cdflag.toString().toLowerCase());
```
Just enforcing consistency in the value here

File: src/utils/dns.js [hunk 22]
```diff
@@ -150,9 +155,9 @@ const performLookup = async (domain, type, endpoint) => {
     };
 };
 
-export const performLookupWithCache = (domain, type, endpoint) => cache(
+export const performLookupWithCache = (domain, type, endpoint, cdflag) => cache(
     performLookup,
-    [ domain, type, endpoint ],
+    [ domain, type, endpoint, cdflag ],
     `dns-${domain}-${type}-${endpoint.endpoint}`,
```
Comment (MattIPv4) [inline src/utils/dns.js]:
```suggestion
    `dns-${domain}-${type}-${endpoint.endpoint}-${cdflag}`,
```
We need the flag in the cache key as it impacts what result we get back

File: src/utils/dig.js [hunk 23]
```diff
@@ -92,6 +98,7 @@ export const parseEmbed = embed => {
         name: descMatch[1],
         type: descMatch[2],
         short: !!descMatch[4],
+        cdflag: descMatch[5],
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
        cdflag: !!descMatch[5],
```
Typecast to a boolean here

File: src/utils/dig.js [hunk 24]
```diff
@@ -5,6 +5,8 @@ import providers from './providers.js';
 import { presentTable } from './table.js';
 import { createEmbed } from './embed.js';
 
+const DNSSEC_DISABLED_WARNING_MESSAGE = '\n\n:warning: cd bit set for request, DNSSEC validation disabled';
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
const DNSSEC_DISABLED_WARNING_MESSAGE = ':warning: cd bit set for request, DNSSEC validation disabled';
```
I take back what I said about adding the newlines here, sorry (see  below)

File: src/utils/dig.js (hunk 20 above)
Comment (MattIPv4) [reply src/utils/dig.js]: Can we also move this value down to when we start using it? Line 70 ish

File: src/utils/dig.js [hunk 25]
```diff
@@ -28,23 +30,27 @@ export const validateDomain = (input, response) => {
...
         // Error message
         if (typeof data === 'object' && data.message)
             return `${digCmd}\n${data.message}`;
 
         // No results
         if (typeof data !== 'object' || !Array.isArray(data.answer) || data.answer.length === 0)
-            return `${digCmd}\nNo records found`;
+            return `${digCmd}\nNo records found${(cdflag
+                ? DNSSEC_DISABLED_WARNING_MESSAGE
+                : '')}`;
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
            return `${digCmd}\nNo records found${cdflag
                ? `\n\n${DNSSEC_DISABLED_WARNING_MESSAGE}`
                : ''}`;
```
Double linebreak make sense for formatting here

File: src/utils/dig.js [hunk 26]
```diff
@@ -63,12 +69,12 @@ export const handleDig = async ({ domain, types, short, provider }) => {
...
         // Keep adding rows until we reach Discord 4096 char limit
         for (const row of sourceRows) {
-            if (output([...finalRows, row]).length > 4096) break;
+            if (output([...finalRows, row]).length > max_length) break;
             finalRows.push(row);
         }
 
         // Render and return final rows
-        return output(finalRows);
+        return output(finalRows) + (cdflag ? DNSSEC_DISABLED_WARNING_MESSAGE : '');
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
        return `${output(finalRows)}${cdflag
          ? `\n${DNSSEC_DISABLED_WARNING_MESSAGE}`
          : ''}`;
```
Single linebreak here as the codeblock for results already has some padding at the bottom

**Review by MattIPv4 — CHANGES_REQUESTED** [review]

File: src/utils/dig.js (hunk 24 above)
Comment (Dhravya) [reply src/utils/dig.js]: It's fine haha. I had it that way before in case different formatting is needed alter on

**Comment by Dhravya** [comment]
Made the changes

File: src/utils/dig.js [hunk 27]
```diff
@@ -61,14 +65,18 @@ export const handleDig = async ({ domain, types, short, provider }) => {
...
-            if (output([...finalRows, row]).length > 4096) break;
+            if (output([...finalRows, row]).length > maxLength) break;
             finalRows.push(row);
         }
 
         // Render and return final rows
-        return output(finalRows);
+        return `${output(finalRows)}${cdflag
+          ? `\n${DNSSEC_DISABLED_WARNING_MESSAGE}`
+          : ''}`;
```
Comment (MattIPv4) [inline src/utils/dig.js]:
```suggestion
            ? `\n${DNSSEC_DISABLED_WARNING_MESSAGE}`
            : ''}`;
```
My bad, was on two-space auto-pilot

**Review by MattIPv4 — APPROVED** [review]
LGTM, thank you for working on this and seeing it through 💙

**Comment by Dhravya** [comment]
cool, thanks a lot! I genuinely learnt so much through this one PR and I definitely look forward to contributing more to this repository!
